### PR TITLE
Multivalue HNSW - implement batch iterator

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR includes the implementation of the HNSW-multi batch-iterator. 

**Main objects this PR modified**
1. `hnsw_batch_iterator` is now a base for `hnsw_single_batch_iterator` and `hnsw_multi_batch_iterator`.
2. moved the creation of batch iterator from `hnsw_factory.cpp` to the derived classes hnsw_single and hnsw_multi.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                                    this->top_candidates_extras.top().second);
+        }
+        this->top_candidates_extras.pop();
+    }
+}
+
+template <typename DataType, typename DistType>
+void HNSWMulti_BatchIterator<DataType, DistType>::updateHeaps(
+    candidatesLabelsMaxHeap<DistType> *top_candidates, DistType dist, idType id) {
+    if (this->lower_bound > dist || top_candidates->size() < this->ef) {
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h:71]: consider changing te ifs order
(first check cand.size < ef, else check the lower_bound > dist)

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h [hunk 2]
```diff
@@ -12,34 +12,38 @@ using spaces::dist_func_t;
...
     tag_t visited_tag;                 // Used to mark nodes that were scanned.
     idType entry_point;                // Internal id of the node to begin the scan from.
     bool depleted;
-    size_t orig_ef_runtime; // Original default parameter to reproduce.
+    size_t ef; // EF Runtime value for this query.
 
     // Data structure that holds the search state between iterations.
-    using candidatesMinHeap = vecsim_stl::min_priority_queue<DistType, idType>;
-    using candidatesMaxHeap = vecsim_stl::max_priority_queue<DistType, idType>;
+    template <typename Value>
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:26]: i find value a bit confusing as a name to the identifier (I would expect that the distance will be called "the value")

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 3]
```diff
@@ -98,13 +98,4 @@ size_t EstimateElementSize(const HNSWParams *params) {
      */
     return size_meta_data + size_total_data_per_element;
 }
-
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_factory.cpp:101]: remove #include "VecSim/algorithms/hnsw/hnsw_batch_iterator.h"

File: src/VecSim/algorithms/hnsw/hnsw_multi.h [hunk 4]
```diff
@@ -45,7 +41,14 @@ class HNSWIndex_Multi : public HNSWIndex<DataType, DistType> {
 
     ~HNSWIndex_Multi() {}
 
+    inline candidatesLabelsMaxHeap<DistType> *getNewMaxPriorityQueue() const override {
+        return new (this->allocator)
+            vecsim_stl::updatable_max_heap<DistType, labelType>(this->allocator);
+    }
+
     inline size_t indexLabelCount() const override;
+    VecSimBatchIterator *newBatchIterator(const void *queryBlob,
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_multi.h:50]: const?

File: src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        this->top_candidates_extras.emplace(top_candidates->top().first,
+                                            top_candidates->top().second); // (distance, label)
+        top_candidates->pop();
+    }
+    rl.results = array_new_len<VecSimQueryResult>(top_candidates->size(), top_candidates->size());
+    // Return results from the top candidates heap, put them in reverse order in the batch results
+    // array.
+    for (int i = (int)(top_candidates->size() - 1); i >= 0; i--) {
+        labelType label = top_candidates->top().second;
+        // TODO: get best score by only checking unvisited vectors under this label.
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h:46]: what about the todo?

File: src/VecSim/algorithms/hnsw/hnsw_single.h [hunk 6]
```diff
@@ -39,7 +35,14 @@ class HNSWIndex_Single : public HNSWIndex<DataType, DistType> {
 
     ~HNSWIndex_Single() {}
 
+    inline candidatesLabelsMaxHeap<DistType> *getNewMaxPriorityQueue() const override {
+        return new (this->allocator)
+            vecsim_stl::max_priority_queue<DistType, labelType>(this->allocator);
+    }
+
     inline size_t indexLabelCount() const override;
+    VecSimBatchIterator *newBatchIterator(const void *queryBlob,
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_single.h:44]: const?

File: src/VecSim/algorithms/hnsw/hnsw_single_batch_iterator.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    candidatesLabelsMaxHeap<DistType> *top_candidates) {
+    while (top_candidates->size() < this->ef && !this->top_candidates_extras.empty()) {
+        top_candidates->emplace(this->top_candidates_extras.top().first,
+                                this->top_candidates_extras.top().second);
+        this->top_candidates_extras.pop();
+    }
+}
+
+template <typename DataType, typename DistType>
+void HNSWSingle_BatchIterator<DataType, DistType>::updateHeaps(
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_single_batch_iterator.h:59]: consider changing te ifs order
(first check cand.size < ef, else check the lower_bound > dist)

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h [hunk 8]
```diff
@@ -97,7 +99,7 @@ HNSW_BatchIterator<DataType, DistType>::scanGraph(candidatesMinHeap &candidates,
         this->candidates.empty()) {
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:99]: this one crashes when commenting  this->top_candidates_extras.empty(),(the batch iterator advanced test fails)
but works without this->candidates.empty()

**Review by meiravgri — COMMENTED** [review]
I like that you manage to give the virtual functions general names so that most of the code lies under the base class :)
I wrote some suggestions, see below

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h [hunk 9]
```diff
@@ -48,45 +52,43 @@ class HNSW_BatchIterator : public VecSimBatchIterator {
...
-    for (int i = (int)(top_candidates.size() - 1); i >= 0; i--) {
-        labelType label = this->index->getExternalLabel(top_candidates.top().second);
-        VecSimQueryResult_SetId(rl.results[i], label);
-        VecSimQueryResult_SetScore(rl.results[i], top_candidates.top().first);
-        top_candidates.pop();
+HNSW_BatchIterator<DataType, DistType>::HNSW_BatchIterator(
+    void *query_vector, HNSWIndex<DataType, DistType> *index, VecSimQueryParams *queryParams,
+    std::shared_ptr<VecSimAllocator> allocator)
+    : VecSimBatchIterator(query_vector, queryParams ? queryParams->timeoutCtx : nullptr,
+                          std::move(allocator)),
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:67]: To indicate that we pass the allocator shared ptr by value

File: src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h (hunk 5 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h:46]: Some other day

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h [hunk 10]
```diff
@@ -122,26 +120,16 @@ HNSW_BatchIterator<DataType, DistType>::scanGraph(candidatesMinHeap &candidates,
         idType curr_node_id = candidates.top().second;
         // If the closest element in the candidates set is further than the furthest element in the
         // top candidates set, and we have enough results, we finish the search.
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:122]: right

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h (hunk 8 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:99]: I think there was some corner case that required this check. I don't want to get into this part of the logic

**Review by meiravgri — DISMISSED** [review]

File: src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        top_candidates->pop();
+    }
+    return rl;
+}
+
+template <typename DataType, typename DistType>
+void HNSWMulti_BatchIterator<DataType, DistType>::fillFromExtras(
+    candidatesLabelsMaxHeap<DistType> *top_candidates) {
+    while (top_candidates->size() < this->ef && !this->top_candidates_extras.empty()) {
+        if (returned.find(this->top_candidates_extras.top().second) == returned.end()) {
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h:60]: isn't it a double check? you check it on line 72

Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_multi_batch_iterator.h:60]: the extras heap can contain a returned value, so this test is needed

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**